play/jamesk.py

```diff
@@ -0,0 +1,32 @@
+def knockknock():
+    import time
+    import random
+
+    knockknock.jokes = {
```

Comment:
This is quite interesting  and I don't recall seeing code using function attributes like this. But it makes sense and is a nice way to allow public access to data associated with a function.

The more traditional way to do this is with a module-level constant like `KNOCKKNOCK_JOKES = { .. }`. The advantage there is that the jokes can be explicitly made part of the public API, and the initialization happens only once. In this current implementation the jokes are redefined every time the function runs.

Reply:
I moved this dict to the module level and, of course, it worked.  However, I don't expect to use this dict in any other function in the module, and the dict creation overhead is not significant enough to motivate the refactor, so I moved it back into the function.

it is a philosophical debate, but I like keeping the data related to the function inside the function (as long as it is not needed by anything else).